### DISCUSSION TITLE
readRect() is reading the wrong color

### DISCUSSION THREAD

**Discussion by kitecraft** [Bodmer/TFT_eSPI#1190] [Q&A - General]
Hello.
First, thank you to Bodmer for the fantastic library.

I'm having an issue with pushRect() pushing the wrong color.  I have a very simple example sketch that shows the problem.
TFT_ESPI_VERSION "2.3.66"

`#include <TFT_eSPI.h>
TFT_eSPI tft = TFT_eSPI();
uint16_t* knobBuffer;

#define SLIDER_KNOB_WIDTH 32
#define SLIDER_KNOB_HEIGHT 14

void setup() {
  tft.init();
  tft.setRotation(3);
  //tft.fillScreen(0x738E);
  tft.fillScreen(TFT_LIGHTGREY);
  knobBuffer = (uint16_t*)malloc(((SLIDER_KNOB_WIDTH + 2) * (SLIDER_KNOB_HEIGHT + 2)) * 2);
}

void loop() {
  int x = random(300);
  int y = random(200);
  tft.readRect(x, y, SLIDER_KNOB_WIDTH, SLIDER_KNOB_HEIGHT, knobBuffer);
  tft.pushRect(x, y, SLIDER_KNOB_WIDTH, SLIDER_KNOB_HEIGHT, knobBuffer);

  delay(3000);
}`

I see this on multiple displays.
Attached is an image of the display showing the mismatched color.  And the script and user_setup.h
I'm curious if this is a buffer size problem?  or maybe a bug was introduced?  or something else? 

Thanks for reading.
Cheers,
Kitecraft.

[tft_pushRectTest.zip](https://github.com/Bodmer/TFT_eSPI/files/6481796/tft_pushRectTest.zip)

**Comment by Bodmer** [comment]
I have run a test with an ESP32 and ST7789 and it runs fine.

To test your setup I need full details of your hardware and software environment (versions etc).

**Comment by kitecraft** [comment]
HI.  Thanks for the response.  Sorry, I should have included this info in the first post.  
I'm using an ESP32 devkit (https://i0.wp.com/microcontrollerslab.com/wp-content/uploads/2019/02/ESP32-pinout-mapping.png?ssl=1), a basic 2.8" ILI9341 touch display, Arduino 1.8.15 and TFT_eSPI version 2.3.66
I don't have any other types of display to try this on otherwise I would.

Thank you.

**Reply by kitecraft** [reply]
Just checked the AnimatedDial example and it's working fine.  btw, that's very helpful example.

**Reply by kitecraft** [reply]
Hi Bodmer
So, I did a bit of digging.
Assumption:  The value I use for fillScreen() is the value of color on line 1341of TFT_eSPI.cpp (readRect() spi mode)

Results:
fillScreen(TFT_YELLOW) = 0xFFE0
line 1341:  color = FFE0
Display looks good


fillScreen(TFT_DARKGREEN) = 0x03E0
line 1341:  color = 03E0
Display looks good

fillScreen(TFT_SKYBLUE) = 0x867D
line 1341:  color = C67D
Display looks wrong

fillScreen(0x5AEB)
line 1341:  color = 7BEF
Display looks wrong

fillScreen(0x9CD3)
line 1341:  color = DEFB
Display looks wrong

I have no idea if it's the my ILI9341 is returning the right value for each of the tft_Read_8() calls though.  
Hope that helps. 
Thanks again.

**Reply by kitecraft** [reply]
Any further info?

**Reply by kitecraft** [reply]
It's these displays:   https://www.amazon.ca/gp/product/B087C3PP9G/ref=ppx_yo_dt_b_asin_image_o01_s00?ie=UTF8&psc=1

**Reply by kitecraft** [reply]
Disappointed at the lack of interest and feedback.  Alas, readrect() just doest seem to work well or consistently.  Oh well.

**Comment by iw7dmh** [comment]
Hi,

it seems I have the same problem here.
I don't know what is wrong with my code. Can you help me please?
My hardware is fairly standard: 5' SSD1963 display - 8 bit parallel - ESP32 DOIT V1

Thank you very much

```
#include <TFT_eSPI.h>
#include <SPI.h>

const int N_SAMPLES = 600;
uint16_t p_rect[N_SAMPLES * 10];

TFT_eSPI tft = TFT_eSPI();

void setup() {
  tft.init();
  tft.setRotation(1);
  delay(1000);
  tft.fillScreen(TFT_BLACK);
  tft.fillRectVGradient(0, 256, N_SAMPLES, 100, TFT_RED, TFT_YELLOW);
  tft.readRect(0, 256+10 , N_SAMPLES, 10, p_rect);  //Read 10 waterfal lines
  tft.pushRect(0, 256+100 + 10 , N_SAMPLES, 10, p_rect);  //Write the same lines below
}

void loop() {
}
```

![image](https://user-images.githubusercontent.com/87143145/200326825-91e0a294-c2c0-40f5-9776-9eb636e149b0.png)

**Reply by Bodmer** [reply]
Run the TFT)ReadWrite_Test sketch and see if the correct pixel values are reported. Post a short section of the results.

Run the Read_User_Setup sketch and post the output here.

**Reply by iw7dmh** [reply]
Hi,
here are the two dumps:

**TFT_ReadWrite_Test**
```
 Pixel value written = 1
 Pixel value read    = 2945
 ERROR                 ^^^^
 Pixel value written = 2
 Pixel value read    = 2945
 ERROR                 ^^^^
 Pixel value written = 4
 Pixel value read    = 2945
 ERROR                 ^^^^
 Pixel value written = 8
 Pixel value read    = 2945
 ERROR                 ^^^^
 Pixel value written = 10
 Pixel value read    = 2945
 ERROR                 ^^^^
 Pixel value written = 20
 Pixel value read    = 2945
 ERROR                 ^^^^
 Pixel value written = 40
 Pixel value read    = 2945
 ERROR                 ^^^^
 Pixel value written = 80
 Pixel value read    = 2945
 ERROR                 ^^^^
 Pixel value written = 100
 Pixel value read    = 2945
 ERROR                 ^^^^
 Pixel value written = 200
 Pixel value read    = 2945
 ERROR                 ^^^^
 Pixel value written = 400
 Pixel value read    = 2945
 ERROR                 ^^^^
 Pixel value written = 800
 Pixel value read    = 2945
 ERROR                 ^^^^
 Pixel value written = 1000
 Pixel value read    = 2945
 ERROR                 ^^^^
 Pixel value written = 2000
 Pixel value read    = 2945
 ERROR                 ^^^^
 Pixel value written = 4000
 Pixel value read    = 2945
 ERROR                 ^^^^
 Pixel value written = 8000
 Pixel value read    = 2945
 ERROR                 ^^^^
```

**Read_User_Setup**

```
TFT_eSPI ver = 2.4.78
Processor    = ESP32
Frequency    = 240MHz
Transactions = Yes
Interface    = Parallel
Display driver = 1963
Display width  = 480
Display height = 800

TFT_CS   = GPIO 33
TFT_DC   = GPIO 15
TFT_RST  = GPIO 32
TFT_WR   = GPIO 4
TFT_RD   = GPIO 2

TFT_D0   = GPIO 12
TFT_D1   = GPIO 13
TFT_D2   = GPIO 26
TFT_D3   = GPIO 25
TFT_D4   = GPIO 17
TFT_D5   = GPIO 16
TFT_D6   = GPIO 27
TFT_D7   = GPIO 14

Font GLCD   loaded
Font 2      loaded
Font 4      loaded
Font 6      loaded
Font 7      loaded
Font 8      loaded
Smooth font enabled
```

**Reply by Bodmer** [reply]
Read is not working. Check your connections to the display, in particular make sure the RD signal is connected.

Some displays do not support reads (unidirectional buffer fitted). Post a link to the vendors page and post a picture of the back of the display board.

**Comment by iw7dmh** [comment]
Hi,

here is the link to the display vendor [LCD 5“SSD1963 TFT](https://www.buydisplay.com/lcd-5-ssd1963-tft-module-touch-screen-display-800x480-mcu-arduino) - The unit is configured as follow: 3.3V - 8bit parallel and GSL1680 capacitive touchscreen. Except from readrect() it works very well.
I still haven't checked the connection between the display RD pin and the ESP32 GPIO-2 pin. I'll provide another feedback asap.
Thank you veru much

**Reply by Bodmer** [reply]
That display supports reads.

**Reply by iw7dmh** [reply]
Hi,

I checked the connection between GPIO2 and RD pin in display connector and it is fine. 
Unless I received a faulted unit I had to investigate in something else.
Anyway thank you very much for your help, I learned how to use some of the **Test and diagnostics** examples.

Best regards

**Comment by keyboardpartner** [comment]
I had the same problem with ILI9341 TFT. Colors read were too bright.
Setting
SPI_READ_FREQUENCY=10000000 (instead of 20000000)
in user setup fixed it.

Thanks again Bodmer for the great library.
